lib/launcher/src/Cardano/Launcher/Node.hs

```diff
@@ -108,38 +119,41 @@ newtype NodePort = NodePort { unNodePort :: Int }
 
 -- | A subset of the @cardano-node@ CLI parameters, used for starting the
 -- backend.
-data CardanoNodeConfig = CardanoNodeConfig
-    { nodeDir             :: FilePath
-    , nodeConfigFile      :: FilePath
-    , nodeTopologyFile    :: FilePath
-    , nodeDatabaseDir     :: FilePath
-    , nodeDlgCertFile     :: Maybe FilePath
-    , nodeSignKeyFile     :: Maybe FilePath
-    , nodeOpCertFile      :: Maybe FilePath
-    , nodeKesKeyFile      :: Maybe FilePath
-    , nodeVrfKeyFile      :: Maybe FilePath
-    , nodePort            :: Maybe NodePort
+data CardanoNodeConfig d = CardanoNodeConfig
+    { nodeDir :: FilePath
+    , nodeConfigFile :: FilePath
+    , nodeTopologyFile :: FilePath
+    , nodeDatabaseDir :: FilePath
+    , nodeDlgCertFile :: Maybe FilePath
+    , nodeSignKeyFile :: Maybe FilePath
+    , nodeOpCertFile :: Maybe FilePath
+    , nodeKesKeyFile :: Maybe FilePath
+    , nodeVrfKeyFile :: Maybe FilePath
+    , nodePort :: Maybe NodePort
     , nodeLoggingHostname :: Maybe String
-    , nodeExecutable      :: Maybe FilePath
-    , nodeOutputFile      :: Maybe FilePath
-    } deriving (Show, Eq)
+    , nodeExecutable :: Maybe FilePath
+    , nodeOutputFile :: Maybe FilePath
+    , nodeSocketPathFile :: MaybeK FilePath d
+    }
+    deriving (Show, Eq)
 
 -- | Spawns a @cardano-node@ process.
 --
 -- IMPORTANT: @cardano-node@ must be available on the current path.
 withCardanoNode
     :: Tracer IO LauncherLog
     -- ^ Trace for subprocess control logging
-    -> CardanoNodeConfig
-    -> (CardanoNodeConn -> IO a)
+    -> CardanoNodeConfig d
+    -- ^ Configuration for the node
+    -> (MaybeK CardanoNodeConn d -> IO a)
```

Comment:
I wanted to avoid the unused socket. I still think it's better to have only one open and  be the relay one. So the fact that they are constraint in length is not relevant. We can have separate paths for the 2 cases (with and without socket), but I think this will create some code duplication.

Reply:
So the issue is just how to ensure the presence/absence of the socket in `CardanoNodeConn` given the configuration that requires it.

Reply:
> I wanted to avoid the unused socket. I still think it's better to have only one open and be the relay one. […] We can have separate paths for the 2 cases (with and without socket), but I think this will create some code duplication.

Yeah, since the socket is a resource, we may want to conserve it. 🤔 The socket is relevant for the node-to-client protocol only, the node-to-node protocol uses TCP/IP, so the node is open for communication even if the socket file has not used.

> We can have separate paths for the 2 cases (with and without socket), but I think this will create some code duplication.

Yeah. I prefer `MaybeK` over the duplication. Another option would be to use a plain `Maybe CardanoNodeConn` and accept the potential proof obligation at run-time, but `MaybeK` provides better value at very small cost, so I prefer that as well.

Alright, convinced, `MaybeK` it is.